### DISCUSSION TITLE
the formats of root CA public key and authentication failure

### DISCUSSION THREAD

**Discussion by TomHsiung** [strongswan/strongswan#1250] [General]
Hello, guys

I have a strongswan server with pem format CA public key. Recently I have to use Apple configurator 2 to generate VPN profiles for my devices. However, the Apple configurator 2 does not accept pem certificates, so switch the pem CA certificate into cer format. After that, I found my device got auth failure when connect my strongswan server.

Does this mean I have to scratch from the ground to re-generate all keys in new format which Apple configurator 2 accepts?

Thanks.

**Comment by tobiasbrunner** [comment]
> so switch the pem CA certificate into cer format

How exactly did you do that?

**Comment by TomHsiung** [comment]
`How exactly did you do that?`

Hi, Tobia

On Mac it is a simple task. First double-clicking the pem CA public key and it will be installed into the app Keychain Access. Later, in the Keychain Access app, you can choose the very pem CA public key and export it into a file with different format.

**Reply by tobiasbrunner** [reply]
Did you check if it's actually the same certificate? Because PEM and DER are simply different encodings (the latter binary, the former a Base64 text representation). There is no reason for an authentication failure unless the certificate is not the correct one or there is a problem with your PKI/certificate in the first place. You might want to provide more information about your certificates and the authentication failure (see #196).

**Reply by TomHsiung** [reply]
Hmmmmmmm, probably it is the problem of the exported CA certificate.

**Comment by TomHsiung** [comment]
> Aug 26 14:14:42 machine charon: 11[NET] received packet: from A[500] to S[500] (316 bytes)
> Aug 26 14:14:42 machine charon: 11[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(REDIR_SUP) N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) ]
> Aug 26 14:14:43 machine charon: 13[MGR] ignoring request with ID 0, already processing
> Aug 26 14:14:43 machine charon: 11[IKE] A is initiating an IKE_SA
> Aug 26 14:14:43 machine charon: 11[CFG] selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_512/ECP_521
> Aug 26 14:14:43 machine charon: 11[IKE] remote host is behind NAT
> Aug 26 14:14:43 machine charon: 11[IKE] sending cert request for "CN=VPN 1"
> Aug 26 14:14:43 machine charon: 11[IKE] sending cert request for "CN=VPN 1"
> Aug 26 14:14:43 machine charon: 11[IKE] sending cert request for "CN=VPN 2"
> Aug 26 14:14:43 machine charon: 11[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(CHDLESS_SUP) N(MULT_AUTH) ]
> Aug 26 14:14:43 machine charon: 11[NET] sending packet: from S[500] to A[500] (389 bytes)
> Aug 26 14:14:43 machine charon: 05[NET] received packet: from A[4500] to S[4500] (376 bytes)
> Aug 26 14:14:43 machine charon: 05[ENC] unknown attribute type INTERNAL_DNS_DOMAIN
> Aug 26 14:14:43 machine charon: 05[ENC] parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) CERTREQ IDr CPRQ(ADDR MASK DHCP DNS ADDR6 DHCP6 DNS6 DOMAIN) N(ESP_TFC_PAD_N) N(NON_FIRST_FRAG) SA TSi TSr N(MOBIKE_SUP) ]
> Aug 26 14:14:43 machine charon: 05[IKE] received cert request for "CN=VPN 2"
> Aug 26 14:14:43 machine charon: 05[CFG] looking for peer configs matching S[sss.sss.sss]...A[vpncomputer]
> Aug 26 14:14:43 machine charon: 05[CFG] selected peer config 'roadwarrior'
> Aug 26 14:14:43 machine charon: 05[IKE] initiating EAP_IDENTITY method (id 0x00)
> Aug 26 14:14:43 machine charon: 05[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
> Aug 26 14:14:43 machine charon: 05[IKE] peer supports MOBIKE
> Aug 26 14:14:43 machine charon: 05[IKE] authentication of 'sss.sss.sss' (myself) with ECDSA-521 signature successful
> Aug 26 14:14:43 machine charon: 05[IKE] sending end entity cert "CN=sss.sss.sss"
> Aug 26 14:14:43 machine charon: 05[ENC] generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
> Aug 26 14:14:43 machine charon: 05[NET] sending packet: from S[4500] to A[4500] (805 bytes)

It stopped at 

> Aug 26 14:14:43 machine charon: 05[NET] sending packet: from S[4500] to A[4500] (805 bytes)

**Reply by tobiasbrunner** [reply]
Then the client probably doesn't like the server certificate sent in the CERT payload there. Either it doesn't support ECDSA, can't match the server identity against a subjectAltName, or it doesn't trust the issuer certificate.

**Reply by TomHsiung** [reply]
No. It worked very well with ECDSA. It just gets trouble when I try to configure the VPN via Jamf Pro. Maybe Jamf Pro does not accept self-signed root CA.

**Reply by tobiasbrunner** [reply]
No idea what that is. But root CAs are always self-signed.

**Comment by TomHsiung** [comment]
Just created new root CA, keys, certs in format of DER. The issue persists with same logs. It seems something is wrong with
>  charon: 05[ENC] generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]

**Comment by TomHsiung** [comment]
Finally figured it out. It seems that a third-party CA must be used to sign certificates. A CA generated by user from IPSEC PKI is not accept by the Jamf Pro service provider.

Thanks.
